### ISSUE TITLE
Apparently missing data for a few hours

### ISSUE THREAD

**Issue by khorton** [SCHibbard/enphase-dashboard#9]
Today's Output vs Time plots are interesting.  There is an about 5.5 hour long period this morning where there was apparently no data returned by the Enphase system.  I come to this conclusion because I see 5.5 hour long straight lines for all panels, with no data points in between the ends of the lines.

Questions: 

1. Is there any other possible cause for this?
2. Does Enphase provide any way to retrieve old data? 
3. If old data can be retrieved, it might be interesting to have the Enphase-Dashboard detect that there are gaps in the data and make an effort to fill those gaps once Enphase starts reporting again

![Screenshot 2024-07-12 at 19 40 08](https://github.com/user-attachments/assets/73705b52-4850-46a0-818d-88db7927ccbe)

**Comment by SCHibbard** [comment]
Yup, it's a data gap.  Assume all other inputs (Powerwall, weather ...) were getting data OK?  Since enphase-dashboard is using the same docker containers the rest of the dashboard is using, hard to figure why it would go down without the others going down.  If you go to your Enphase account online does it show data gaps for that time period as well? I have never had data loss problems with the Enphase data, especially since it is pulled from the local network.  I HAVE had data gaps with the hourly pricing data I get from our electric utility, and wrote a script like what you described ... look for gaps and fill them.  History is not stored locally in the Enphase inverters, but IS available with the Enphase online API.  I recall there was a free level for API access, but enough calls / time period to probably get the history needed.  I'll look into it later, but it will be a project!

**Comment by khorton** [comment]
Powerwall and weather had data through the whole day.  The Energy view from Enphase online shows data for every interval, and it has the typical shape I see in the morning.

Puzzling, but not a big deal, so not worthwhile investing a lot of effort into unless other people are reporting it frequently.

**Comment by SCHibbard** [comment]
If you get a chance to look online at Enphase, (Energy, then pick day). If the missing data shows up there it's most likely a problem in telegram/Influx.  If not, the Enphase Gateway was not outputting data. Thanks!

**Comment by khorton** [comment]
Looking at the pypowerwall logs, I see short periods on many days with entries like:

`07/13/2024 01:12:13 AM [pypowerwall] [ERROR] Failed to get /api/system_status/grid_status`

and which is often followed by:

`07/13/2024 01:12:13 AM [urllib3.connectionpool] [WARNING] Connection pool is full, discarding connection: 192.168.0.184. Connection pool size: 15`

I know nothing on this subject, but my first guess was that I should set a larger connection pool for urllib3.  But online research suggests this log item may be expected, so I'm not changing anything yet

**Comment by khorton** [comment]
And, the URL in the urllib3.connectionpool is for the Tesla Gateway, so nothing to do with Enphase data.  False alarm

**Comment by SCHibbard** [comment]
Interesting, are the log entries the same times as your inverter data gaps?

**Comment by khorton** [comment]
No, the log entries don't align with the inverter data gaps.  Some days only have a couple `[pypowerwall] [ERROR] Failed to get /api/system_status/grid_status` log entries, other days have many dozen.

**Comment by SCHibbard** [comment]
Did some investigation into this ... not good.  Was easy to make a script to find all gaps in InfluxDB inverter data.  Did that, then looked into Enphase's current online API (https://developer-v4.enphase.com/).  Should have looked first, then coded. Enphase is trying to make the API a good revenue generator.  They have not designed the API for personal use, but instead commercial developer use. The free plan only gives you site data, not device (individual inverter) data.  To get inverter production, you need at least the Kw plan, which is $250/month. Seems a bit extreme to charge $250/mo to get your own data, but I see no other way.  So no solution at the moment.

**Comment by khorton** [comment]
I had another event today where there was no Enphase panel data for about 2.5 hours from about 13:05 to 15:35 EDT. All Powerwall data was good during that period, so it seems to be either Enphase issues, or a glitch in the data recording.

The Enphase data at https://enlighten.enphaseenergy.com looks normal - i.e. you can see the solar energy varying throughout that 2.5 hour period.

The telegraf logs have entries like the following every 5 minutes during that 2.5 hour period:

```2025-06-10T17:10:03Z E! [inputs.http] Error in plugin: [url=https://192.168.0.185/api/v1/production/inverters]: Get "https://192.168.0.185/api/v1/production/inverters": dial tcp 192.168.0.185:443: connect: no route to host```

Looks like either Enphase was down, or there was some internet issue.

I'm wondering if you had the same issue, or if it was more localized.

**Comment by SCHibbard** [comment]
I see no loss of data for the recent past, including yesterday at noon-15:00 CT.  The Telegraf error looks like network problems but it is interesting you say there is no loss of data at the Enphase Cloud. That would mean either they buffer data for Cloud upload, or the network connection worked to them but not local.  I doubt the later, so it may be the former ... just don't know how to figure that out. Maybe a post on the Enphase board to see if anyone knows?
